Add support for inline sources

cpuguy83 opened the pull request:
This allows you to specify files or trees inline in yaml spec.

Example:

```yaml
sources:
  someFile:
    inline:
      file:
      contents: "hello!"
  someDir:
    inline:
      dir:
        files:
          foo:
            contents: "I am foo!"
```

You can also set permissions and ownership.

pmengelbert commented on spec.go:
```diff
@@ -225,6 +225,60 @@ type SourceBuild struct {
+	GID int `yaml:"gid,omitempty" json:"gid,omitempty"`
+}
+
+// SourceInlineDir is used by by [SourceInline] to represent a filesystem directory.
+type SourceInlineDir struct {
+	// Files is the list of files to include in the directory.
+	// The map key is the name of the file.
+	//
+	// Files with path separators in the key will be rejected.
+	//
+	// Files are created after directories are created.
+	Files map[string]*SourceInlineFile `yaml:"files,omitempty" json:"files,omitempty"`
+	// Dirs is the list of directories to include in the directory.
+	// The map key is the name of the directory.
+	//
+	// Directories with path separators in the key will be rejected.
+	//
+	// Directories are created before files are created.
+	Dirs map[string]*SourceInlineDir `yaml:"dirs,omitempty" json:"dirs,omitempty"`
+	// Permissions is the octal permissions to set on the directory.
+	Permissions fs.FileMode `yaml:"permissions,omitempty" json:"permissions,omitempty"`
+
+	// UID is the user ID to set on the directory and all files and directories within it.
+	// UID must be greater than 0
+	UID int `yaml:"uid,omitempty" json:"uid,omitempty"`
+	// GID is the group ID to set on the directory and all files and directories within it.
+	// GID must be greater than 0
+	GID int `yaml:"gid,omitempty" json:"gid,omitempty"`
+}
```
Inline files make a lot of sense for one-off things that don't deserve their own file, but I don't know if I would want to allow directories. It seems unnecessary to allow complex nested inlining when the filesystem is right there and available as a `SourceTypeContext`.

I think it could get quickly get very indented and cluttered. This functionality is already available by other means, for example a `Build` source type with a local dir provided as a context and a small inline dockerfile to set the perms and uid/gid of the files/directories, which can be more concise due to `(chown|chmod) -R`.

cpuguy83 replied:
Context is not really the same thing.
Context is client-side, different clients have different contexts.

Using `Build` is also significantly more complex (on the backend) and requires a remote image to do anything at all.

I thought I put this in the commit but apparently I did not: The main intent for these is to simplify testing and not require tests to hit registries.

pmengelbert replied:
> Context is not really the same thing. Context is client-side, different clients have different contexts.
> 
Understood

> Using `Build` is also significantly more complex (on the backend) and requires a remote image to do anything at all.
> 
Make sense

> I thought I put this in the commit but apparently I did not: The main intent for these is to simplify testing and not require tests to hit registries.
Sounds good. Was wondering what the reason was for this.

pmengelbert replied:
> Context is not really the same thing. Context is client-side, different clients have different contexts.
> 

So actually I'm not sure I fully understand this. You mean that a context can be all kinds of things, depending on the client? For example, with `buildctl` I can do this:

```bash
buildctl build \                                                                                       --frontend=dockerfile.v0 \                                                                     --local context=. \                                                                            --oci-layout=foo=./oci \                                                                       --opt context:bar=oci-layout:foo@sha256:760c7b6<...> \                                         --local dockerfile=. \
        --output=type=local,dest=../hello

# Dockerfile
FROM bar
```

An in this case, the directory `oci` is not actually the context (because it's json blobs / tarballs of the oci layout dir), the actual context is the fs of the image *represented* by the `oci` layout dir.

Is that right?

cpuguy83 replied:
I mean a context is dependent on the build invocation.
The client tells you what the context is and is not declarative.

adamperlin commented on files.go:
```diff
@@ -0,0 +1,184 @@
+	return func(st llb.State) llb.State {
+		perms := f.Permissions.Perm()
+		if perms == 0 {
+			perms = 0o644
```
Maybe add this as a const?

cpuguy83 replied:
Done

adamperlin commented on load.go:
```diff
@@ -365,14 +371,17 @@ func (s *Spec) FillDefaults() {
 
 func (s Spec) Validate() error {
 	for name, src := range s.Sources {
+		if strings.ContainsRune(name, os.PathSeparator) {
```
On this note, have we specifically defined what are and are not valid source names?

cpuguy83 replied:
We have not, but should indeed.

adamperlin commented on spec.go:
```diff
@@ -225,6 +225,60 @@ type SourceBuild struct {
+	// Permissions is the octal file permissions to set on the file.
+	Permissions fs.FileMode `yaml:"permissions,omitempty" json:"permissions,omitempty"`
+	// UID is the user ID to set on the directory and all files and directories within it.
+	// UID must be greater than 0
```
It looks like in the validation code we're just testing `< 0`, could we clarify this comment vs. the validation logic?

cpuguy83 replied:
Fixed

adamperlin commented on spec.go:
```diff
@@ -225,6 +225,60 @@ type SourceBuild struct {
+
+	// UID is the user ID to set on the directory and all files and directories within it.
+	// UID must be greater than 0
+	UID int `yaml:"uid,omitempty" json:"uid,omitempty"`
```
Same as above, just making sure comment matches validation

cpuguy83 replied:
Nice catch. I think this was copilot :)
It tries hard.

cpuguy83 commented:
Also realized we can drop the "Inline" field in the SourceBuild struct.

adamperlin commented on files.go:
```diff
@@ -0,0 +1,189 @@
+	if s.Dir != nil {
+		s.Dir.Doc(w, name)
+	}
+
```
nit: extra newline

adamperlin commented on frontend/debug/handle_sources.go:
```diff
@@ -19,14 +19,15 @@ func HandleSources(ctx context.Context, gwc client.Client, spec *dalec.Spec) (cl
 		}
 
-		to := llb.Scratch().File(llb.Mkdir("/"+name, 0o755, llb.WithParents(true)))
-		sources = append(sources, to.File(llb.Copy(st, "/", "/"+name, dalec.WithDirContentsOnly())))
```
Was this copy operation not necessary?

cpuguy83 replied:
Yes, it was adding an extra nesting unexpectedly.

cpuguy83 commented:
I'm going to move the new test to a unit test.

cpuguy83 commented:
I've updated this to test inline sources via unit tests.

I've also removed support for nested directories.
We can add this back easily enough in the future if needed, but after discussing with @pmengelbert last week I think it is best to keep this simple.
If multiple directories are needed someone can use multiple sources and arrange things as part of the build step.

adamperlin commented:
Wait what was the reasoning for removing inline directories? In my opinion natively supporting that seems simpler than trying to achieve it through multiple sources/using build context. I think nested directories are very useful for testing purposes (I can already think of cases where this would have helped me), and I personally don't feel like it adds that much complication. Yes, it would get deeply nested if you used the feature beyond its intended purpose, but for putting in one or two levels of directory nesting it doesn't seem unreasonable to me. Definitely not a blocker by any means, but just thought I'd add my opinion

cpuguy83 commented:
This is rebased and all green.

cpuguy83 commented:
@adamperlin There is no way to specify nested directories at all.

adamperlin commented:
Ah yes, I see that in the spec for `SourceInlineDir` now

adamperlin reviewed (approved):
Looks good to me!
